prometheus/wrap.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// WrapWith returns a Registerer wrapping the provided Registerer. Collectors
+// registered with the returned Registerer will be registered with the wrapped
+// Registerer in a modified way. The modified Collector adds the provided Labels
+// to all Metrics it collects (as ConstLabels). The Metrics collected by the
+// unmodified Collector must not have any of those labels on their own.
+//
+// WrapWith provides a way to add fixed labels to a subset of Collectors. It
+// should not be used to add fixed labels to all metrics exposed.
+//
+// The Collector example demonstrates a use of WrapWith.
+func WrapWith(labels Labels, reg Registerer) Registerer {
+	return &wrappingRegisterer{
+		wrappedRegisterer: reg,
+		labels:            labels,
+	}
+}
+
+// WrapWithPrefix returns a Registerer wrapping the provided Registerer. Collectors
+// registered with the returned Registerer will be registered with the wrapped
+// Registerer in a modified way. The modified Collector adds the provided prefix
+// to the name of all Metrics it collects.
+//
+// WrapWithPrefix is useful to have one place to prefix all metrics of a
+// sub-system. To make this work, register metrics of the sub-system with the
+// wrapping Registerer returned by WrapWithPrefix.
+func WrapWithPrefix(prefix string, reg Registerer) Registerer {
```

Comment:
Similar suggestion to `WrapWith`, the registerer should be the first argument and the function name I'd suggest to just be `WithPrefix`.